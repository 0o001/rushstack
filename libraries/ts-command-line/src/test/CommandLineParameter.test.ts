// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as colors from 'colors';

import { DynamicCommandLineParser } from '../DynamicCommandLineParser';
import { DynamicCommandLineAction } from '../DynamicCommandLineAction';

describe('CommandLineParameter', () => {
  const commandLineParser: DynamicCommandLineParser = new DynamicCommandLineParser(
    {
      toolFilename: 'example',
      toolDescription: 'An example project'
    }
  );
  commandLineParser.defineFlagParameter({
    parameterLongName: '--global-flag',
    parameterShortName: '-g',
    description: 'A flag that affects all actions'
  });

  const action: DynamicCommandLineAction = new DynamicCommandLineAction({
    actionVerb: 'do-job',
    summary: 'does the job',
    documentation: 'a longer description'
  });
  commandLineParser.addAction(action);

  action.defineChoiceParameter({
    parameterLongName: '--choice',
    parameterShortName: '-c',
    description: 'A choice',
    alternatives: [ 'one', 'two' ],
    defaultValue: 'one'
  });
  action.defineFlagParameter({
    parameterLongName: '--flag',
    parameterShortName: '-f',
    description: 'A flag'
  });
  action.defineIntegerParameter({
    parameterLongName: '--integer',
    parameterShortName: '-i',
    description: 'An integer',
    argumentName: 'NUMBER'
  });
  action.defineStringParameter({
    parameterLongName: '--string',
    parameterShortName: '-s',
    description: 'A string',
    argumentName: 'TEXT'
  });
  action.defineStringListParameter({
    parameterLongName: '--string-list',
    parameterShortName: '-l',
    description: 'A string list',
    argumentName: 'LIST'
  });

  it('prints the global help', () => {
    const helpText: string = colors.stripColors(commandLineParser.renderHelpText());
    expect(helpText).toMatchSnapshot();
  });

  it('prints the action help', () => {
    const helpText: string = colors.stripColors(action.renderHelpText());
    expect(helpText).toMatchSnapshot();
  });
});
