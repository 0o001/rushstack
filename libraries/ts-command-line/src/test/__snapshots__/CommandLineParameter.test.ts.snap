// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandLineParameter prints the action help 1`] = `
"usage: example do-job [-h] [-c {one,two}] [-f] [-i NUMBER] [-s TEXT] [-l LIST]

a longer description

Optional arguments:
  -h, --help            Show this help message and exit.
  -c {one,two}, --choice {one,two}
                        A choice
  -f, --flag            A flag
  -i NUMBER, --integer NUMBER
                        An integer
  -s TEXT, --string TEXT
                        A string
  -l LIST, --string-list LIST
                        A string list
"
`;

exports[`CommandLineParameter prints the global help 1`] = `
"usage: example [-h] [-g] <command> ...

An example project

Positional arguments:
  <command>
    do-job           does the job

Optional arguments:
  -h, --help         Show this help message and exit.
  -g, --global-flag  A flag that affects all actions

For detailed help about a specific command, use: example <command> -h
"
`;
